README.md

```diff
@@ -170,6 +194,95 @@ ThriftService instances are not expected to be useful outside the process of
+
+Since HTTP/1.1 only allows one outstanding request and response on the wire at
+any time, and HTTP/1.1 with the ill-fated "pipeline" mode allows requests and
+responses to be sent FIFO but forbids out-of-order responses, the sequence
+number is useless.
+HTTP/2.0 does support multiplexing, but does not require sequence identifiers
+in the request or response body to do so.
+Despite this folly, HTTP request handlers are obliged to use envelopes and
+match the number of a response to the number of a request, but callers are not
+obliged to choose sensible sequence numbers.
+
+By default, services dispatch on the given function name. Services that
+implement multiple Thrift services have the option of requiring the Thrift
+service name and function name in the message envelope, with an arbitrary
+delimiter, typically a single colon, e.g., "Service:function".
+Over TChannel, this corresponds to the frame's arg1 and is always
+"Service::function", the service name and function name delimited by two
+colons.
+
+The message type is an enumeration of one of 'CALL', 'ONEWAY', 'RESULT', or
+'EXCEPTION'.
+'CALL' and 'ONEWAY' are appropriate for requests, where 'ONEWAY' is for
+functions with the `oneway` Thrift IDL modifier keyword, implying they expect
+no response.
+'RESULT' is for both Thrift success and exception cases.
+For exceptions that cannot be captured by the Thrift result struct, the
+'EXCEPTION' message type implies that the body is a Thrift Exception struct
+instead of the result struct from the IDL.
+
+```thrift
+enum ExceptionType {
+  UNKNOWN = 0
+  UNKNOWN_METHOD = 1
+  INVALID_MESSAGE_TYPE = 2
+  WRONG_METHOD_NAME = 3
+  BAD_SEQUENCE_ID = 4
+  MISSING_RESULT = 5
+  INTERNAL_ERROR = 6
+  PROTOCOL_ERROR = 7
+  INVALID_TRANSFORM = 8
+  INVALID_PROTOCOL = 9
+  UNSUPPORTED_CLIENT_TYPE = 10
+}
+
+struct TApplicationException {
+  1: optional string message
+  2: optional ExceptionType type
+}
+```
+
+These cases are expressed in TChannel instead with an Error frame with the
+corresponding TChannel transport error code and an arbitrary message.
+
+The base message envelope constructor is available as the `thrift.Message`,
+that is also a copy constructor.
+Messages have the following properties:
+
+- `version` defaults to 0. Version 1 has a more elaborate message header that
+  expresses the message envelope version number and provides no additional
+  value. There may or may never be another message envelope version.
+- `type` is one of 'CALL', 'ONEWAY', 'RESULT', or 'EXCEPTION'
+- `id` is a sequence number. Really, it is any number. If you are a client, you
+  can pick. If you are a server, you must respect the client's choice.
+- `body` is either your result object, or an exception object that has a `type`
+  and `message`.  The `type` must be one of `UNKNOWN`, `UNKNOWN_METHOD`,
+  `INVALID_MESSAGETYPE`, `WRONG_METHOD_NAME`, `BAD_SEQUENCE_ID`,
+  `MISSING_RESULT`, `INTERNAL_ERROR`, `PROTOCOL_ERROR`,
```

Comment:
I wouldn't document the payload for the EXCEPTION type here and just say it's the TApplicationException as defined above.